src/browser/providers/tidal.js

```diff
@@ -0,0 +1,266 @@
+        (this.controls.prev && this.controls.prev.click()) || super.previous();
+    }
+
+    play() {
+        (!this.activePlayer.isPlaying() &&
+            this.controls.play &&
+            this.controls.play.click()) ||
+            super.play();
+    }
+
+    pause() {
+        (this.activePlayer.isPlaying() &&
+            this.controls.play &&
+            this.controls.play.click()) ||
+            super.pause();
+    }
+
+    playPause() {
+        (this.controls.play && this.controls.play.click()) || super.playPause();
+    }
+
+    setShuffle(isShuffle) {
+        if (this.controls.shuffle) {
+            if (
+                (!this.isShuffle() && isShuffle) ||
+                (this.isShuffle() && !isShuffle)
+            ) {
+                this.controls.shuffle.click();
+            }
+        }
+    }
+
+    isShuffle() {
+        if (this.controls.shuffle) {
+            const classes = Array.from(this.controls.shuffle.classList);
+            return classes.some(c => c.startsWith('active'));
+        } else {
+            return super.isShuffle();
+        }
+    }
+
+    getLoopStatus() {
+        if (this.controls.repeat) {
+            const classes = Array.from(this.controls.repeat.classList);
+            if (classes.some(c => c.startsWith('once'))) {
+                return LoopStatus.TRACK;
+            } else if (classes.some(c => c.startsWith('all'))) {
+                return LoopStatus.PLAYLIST;
+            } else {
+                return LoopStatus.NONE;
+            }
+        } else {
+            return super.getLoopStatus();
+        }
+    }
+
+    setLoopStatus(status) {
+        if (this.controls.repeat) this.controls.repeat.click();
+        else super.setLoopStatus(status);
+    }
+
+    setVolume(volume) {
+        super.setVolume(volume);
+        if (this.controls.volumeProgress) {
+            this.controls.volumeProgress.value = volume * 100;
+            this.control.volumeProgress.style =
+                '--val:' + volume * 100 + '; --max:100; --min:0;';
+        }
+    }
+
+    getVolume() {
+        if (this.controls.volumeProgress) {
+            return parseInt(this.controls.volumeProgress.value, 10) / 100;
+        }
+        return super.getVolume();
+    }
+
+    setRate() {
+        // disabled
+    }
+}
+
+Playback = TidalPlayback;
+
+window.addEventListener('mpris2-setup', function() {
+    /*
+     * Since tidal is loaded dynamically, we need to wait until the playback controls are added
+     * to the DOM before we can get the elements.
+     *
+     * We create an observer to monitor the added nodes, and set the playback controls when the
+     * player is loaded.
+     *
+     */
+    const playerObserver = new MutationObserver(mutations => {
+        let loaded = false;
+        mutations.forEach(mutation => {
+            mutation.addedNodes.forEach(n => {
+                if (n.className.startsWith('footerPlayer')) {
+                    loaded = true;
+                }
```

Comment:
The if statement is running for all mutations in the dom and for some `className.startsWith` throws a TypeError. ie:
```
Uncaught TypeError: Cannot read property 'startsWith' of undefined
    at tidal.js:170
    at NodeList.forEach (<anonymous>)
    at tidal.js:169
    at Array.forEach (<anonymous>)
    at MutationObserver.<anonymous> (tidal.js:168)
(anonymous)	@	tidal.js:170
(anonymous)	@	tidal.js:169
(anonymous)	@	tidal.js:168
```
A quick fix would be to check if className exists and is of type string. Isn't there a way to limit the observer, tho?

Reply:
Hmm, didn't think about that.
I couldn't find a way to limit it and I think its fine tbh, but I can at least stop the observer when the page has loaded so it doesn't keep running even after the page is loaded and the elements are grabbed.